### DISCUSSION TITLE
There is a report that TST is not compatible with Container Tab Groups

### DISCUSSION THREAD

**Discussion by metastable-void** [piroor/treestyletab#3214] [General]
I am a developer of [Container Tab Groups](https://github.com/menhera-org/TabArray). I saw a comment (https://www.ghacks.net/2022/09/26/container-tab-groups-group-sites-securely-in-firefox/#comment-4549897) describing a breakage of TST when our add-on is enabled.

This add-on sorts tabs according to containers. I made a research on this but as of now yet to find the problem. What sort of breakage is possible and how we can prevent ones? I would like to make the both add-ons play nicely. Thanks.

Link: https://github.com/menhera-org/TabArray/issues/32

**Comment by irvinm** [comment]
@metastable-void for what it is worth, I tried a clean profile and just loaded CTG and TST together and didn't have any problems.  I created new tabs in different containers from both addons, created trees, deleted tabs ... all the while I saw no loss of tree hierarchy.  I did read the comment that user made, but obviously it was a fairly generic comment about his situation.  It might be helpful if you ask them to see if they can reproduce it again and file a defect that we can all inspect.

At the end of the day, TST is driven by the relationship between the tabs and the driver for that is `openerTabId` and as long as that is set where appropriate, the tree hierarchy can be maintained.  https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/tabs/Tab

**Reply by piroor** [reply]
I've seen a breakage with the combination of TST and CTG.

Conditions:

* TST: default options
* CTG: Experimental Features => When an external application opens a tab => Open in last active container

Steps to reproduce:

1. Open a window with a tab with the default contianer.
2. Open TST sidebar panel.
3. Long-press on the "+" button in the sidebar, and choose "Personal". It opens a new tab with "Personal" container.
   Now we have two tabs:
   * A (default)
   * B (Personal)
4. Prepare a bookmark folder containing two bookmarks. I've used `https://example.org/?1` and `https://example.org/?2`.
5. Middle-click on the bookmark folder to open two bookmarks in tabs.

Expected result:

* Two tabs are opened with the "Personal" container.
* Those tabs are grouped under a tab provided by TST, like:
  * A (default)
  * B (Personal)
  * Group tab (what container should be used?)
    * `https://example.org/?1` (Personal)
    * `https://example.org/?2` (Personal)

Actual result:

* A (default)
* Group tab (default)
* B (Personal)
* `https://example.org/?1` (Personal)
  * `https://example.org/?2` (Personal) : this is a child of the group tab (not the preceding tab!)

Hmm, I can't imagine how they should work together on this case...

**Comment by piroor** [comment]
Example possible conflict points I think:

* "Open in New Container Tab" command in the tab context menu. It duplicates the source tab with the specified container and place it as configured, but CTG may close the tab or move the tab to elsewhere before/after TST handles it.
* Long-press on the "+" button => a new tab with container. It opens a new tab with the specified container and place it as configured, but CTG may close the tab or move the tab to elsewhere before/after TST handles it.
* "Open Link in New Container Tab" command in the context menu on a link in a webpage. It opens a tab with `openerTabId` with the specified container and TST attaches the tab to the source tab automatically, but CTG may move the tab to elsewhere before/after TST handles it.
* Dragging a tab and dropping it onto another tab with different container. TST moves such a dragged tab next to the drop target tab and attach it to the target as a child tab, but CTG may move the tab to elsewhere before/after TS T handles it.

The root problem is: everything is asynchronous. `tabs.onCreated`, `tabs.onMoved` and `tabs.onUpdated` are notified to listener addons asynchronously, so TST and CTG will handle a tab at same time in parallel. 

For example, they addons may move a tab to somewhere. Assume that TST moves the tab to the position 4, CTG moves the tab to the position 10. But the tab is just one and it can be placed just one place. If TST won the tab is placed at the positon 4, otherwise CTG won and the tab is placed at the position 10. But they don't think that the `tabs.move()` API call was avoided - even if the tab is placed at 10, TST still think that the tab was placed at 4. Such a mismatching internal status causes breakage.

To work TST and CTG together well, I think that CTG need to handle all tabs at first, and TST should handle the tab after that. For exampe, on TST side:

```javascript
browser.tabs.onCreated.addListener(async tab => {
  const updatedTab = await browser.runtime.sendMessage(CTG_ID, { type: 'wait-until-initialized', tabId: tab.id });
  if (!updatedTab) // If CTG closed the tab, ignore it.
    return;
  tab.index = updatedTab.index;
  tab.cookieStoreId = updatedTab.cookieStoreId;
  ...
});

browser.tabs.onMoved.addListener(async (tabId, moveInfo) => {
  const updatedMoveInfo = await browser.runtime.sendMessage(CTG_ID, { type: 'wait-until-moved', tabId });
  if (!updatedMoveInfo) // If CTG closed the tab, ignore it.
    return;
  moveInfo.toIndex = updatedMoveInfo.toIndex;
  ...
});
```

Of course such `await`s will make everything slow, so TST should `await` only when it detects that CTG is working. TST may need to contact to CTG (started before TST) at first, and need to wait contact from CTG (started after TST).

How about this idea? @metastable-void

**Reply by metastable-void** [reply]
Thanks for the replies. I agree with the point that in the WebExtensions world, explicit cooperation from both add-ons is sometimes needed. I have just created an empty entry point of the APIs for CTG (https://github.com/menhera-org/TabArray/commit/b2f083bfcf56bff88b4071d0f45114132592ca1a). I will implement the needed logic in [branch compat-treestyletab](https://github.com/menhera-org/TabArray/tree/compat-treestyletab).
